src/fsharp/NameResolution.fs

```diff
@@ -3260,38 +3260,45 @@ let rec ResolvePartialLongIdentInModuleOrNamespaceForRecordFields (ncenv: NameRe
     // get record type constructors
     let tycons = 
         mty.TypeDefinitions
-        |> List.filter (fun tcref -> not (tcref.LogicalName.Contains(",")))
-        |> List.filter (fun tycon -> tycon.IsRecordTycon)
-        |> List.filter (fun tycon -> not (IsTyconUnseen ad g ncenv.amap m (modref.NestedTyconRef tycon)))
+        |> List.filter (fun tcref -> 
+               (not (tcref.LogicalName.Contains(","))) &&
+               tcref.IsRecordTycon &&
+               (not (IsTyconUnseen ad g ncenv.amap m (modref.NestedTyconRef tcref))))
 
     let ilTyconNames = 
         mty.TypesByAccessNames.Values
-        |> Seq.toList
-        |> List.choose (fun (tycon:Tycon) -> if tycon.IsILTycon then Some tycon.DisplayName else None)
+        |> Seq.choose (fun (tycon:Tycon) -> if tycon.IsILTycon then Some tycon.DisplayName else None)
         |> Set.ofSeq      
     
     match plid with 
     | [] -> 
-        // Collect up the accessible sub-modules 
-       (mty.ModulesAndNamespacesByDemangledName 
-          |> NameMap.range 
-          |> List.filter (fun x -> x.DemangledModuleOrNamespaceName |> notFakeContainerModule ilTyconNames)
-          |> List.filter (fun x -> x.DemangledModuleOrNamespaceName |> IsInterestingModuleName)
-          |> List.map modref.NestedTyconRef
-          |> List.filter (IsTyconUnseen ad g ncenv.amap m >> not)
-          |> List.filter (EntityRefContainsSomethingAccessible ncenv m ad)
-          |> List.map ItemForModuleOrNamespaceRef)
+        let accessibleSubModules =
+            mty.ModulesAndNamespacesByDemangledName 
+            |> NameMap.range 
+            |> List.choose (fun x ->
+                let demangled = x.DemangledModuleOrNamespaceName
+                if notFakeContainerModule ilTyconNames demangled && IsInterestingModuleName demangled then
+                    let entityRef = modref.NestedTyconRef x
+                    if (IsTyconUnseen ad g ncenv.amap m entityRef |> not) && EntityRefContainsSomethingAccessible ncenv m ad entityRef then
+                        Some (ItemForModuleOrNamespaceRef entityRef)
+                    else
+                        None
+                else
+                    None)
 
-       // Collect all accessible record types
-       @ (tycons |> List.map (modref.NestedTyconRef >> ItemOfTyconRef ncenv m) )
-       @ [ // accessible record fields
-            for tycon in tycons do
-                if IsEntityAccessible ncenv.amap m ad (modref.NestedTyconRef tycon) then
-                    let ttype = FreshenTycon ncenv m (modref.NestedTyconRef tycon)
+        let accessibleRecordFields =
+            [for tycon in tycons do
```

Comment:
What about `List.collect` instead of the list expression? It compiles into a `seq` which is converted into a `list` at end.